Xamarin.PropertyEditing.Tests/PredefinedValuesViewModelTests.cs

```diff
@@ -0,0 +1,221 @@
+			return name;
+		}
+
+		protected string GetNameOrDefault (T value)
+		{
+			return Values.Where (kvp => Equals (value, kvp.Value)).Select (kvp => kvp.Key).FirstOrDefault ();
+		}
+
+		protected override void AugmentPropertyMock (Mock<IPropertyInfo> propertyMock)
+		{
+			var predefined = propertyMock.As<IHavePredefinedValues<T>> ();
+			predefined.SetupGet (h => h.IsConstrainedToPredefined).Returns (IsConstrained);
+			predefined.SetupGet (h => h.PredefinedValues).Returns (Values);
+		}
+	}
+
+	internal abstract class ConstrainedPredefinedValuesViewModelTests<T>
+		: PredefinedValuesViewModelTests<T>
+	{
+		protected override bool IsConstrained => true;
+
+		protected abstract T GetOutOfBoundsValue ();
+
+		protected abstract string GetOutOfBoundsValueName ();
+
+		[Test]
+		public void SetValueOutOfBounds ()
+		{
+			T value = GetOutOfBoundsValue ();
+
+			var vm = GetBasicTestModel ();
+			T originalValue = vm.Value;
+			string originalValueName = vm.ValueName;
+			Assume.That (vm.Value, Is.Not.EqualTo (value));
+
+			bool changed = false;
+			vm.PropertyChanged += (sender, args) => {
+				if (args.PropertyName == nameof(vm.Value))
+					changed = true;
+			};
+
+			vm.Value = value;
+			Assert.That (vm.Value, Is.EqualTo (originalValue));
+			Assert.That (vm.ValueName, Is.EqualTo (originalValueName));
+			Assert.That (changed, Is.False);
+		}
+
+		[Test]
+		public void SetValueNameOutOfBounds ()
+		{
+			string valueName = GetOutOfBoundsValueName ();
+
+			var vm = GetBasicTestModel ();
+			T originalValue = vm.Value;
+			string originalValueName = vm.ValueName;
+			Assume.That (vm.ValueName, Is.Not.EqualTo (valueName));
+
+			bool changed = false;
+			vm.PropertyChanged += (sender, args) => {
+				if (args.PropertyName == nameof(vm.ValueName))
+					changed = true;
+			};
+
+			vm.ValueName = valueName;
+			Assert.That (vm.Value, Is.EqualTo (originalValue));
+			Assert.That (vm.ValueName, Is.EqualTo (originalValueName));
+			Assert.That (changed, Is.False);
+		}
+	}
+
+	internal enum PredefinedEnumTest
+		: int
+	{
+		None = 0,
+		First = 1,
+		Second = 2,
+		Eigth = 8
+	}
+
+	[TestFixture]
+	internal class EnumPredefinedViewModelTests
+		: ConstrainedPredefinedValuesViewModelTests<int>
+	{
+		public EnumPredefinedViewModelTests ()
+		{
+			this.values = (int[]) Enum.GetValues (typeof(PredefinedEnumTest));
+			this.names = Enum.GetNames (typeof(PredefinedEnumTest));
+
+			var v = new Dictionary<string, int> (this.values.Length);
+			for (int i = 0; i < this.values.Length; i++) {
+				v.Add (this.names[i], this.values[i]);
+			}
+
+			this.predefinedValues = v;
+		}
+
+		protected override IReadOnlyDictionary<string, int> Values => this.predefinedValues;
+
+		protected override string GetOutOfBoundsValueName ()
+		{
+			return "foo";
+		}
+
+		protected override int GetOutOfBoundsValue ()
+		{
+			return -1;
+		}
+
+		protected override int GetRandomTestValue (Random rand)
+		{
+			int index = rand.Next (0, this.values.Length);
```

Comment:
It is a fixed seed, although test values obviously may shift as tests are added. All of the value based tests already run off this, this is just an override specific to the new interface ;).